### PR TITLE
wallet: simplify and batch zap wallet txes process

### PR DESCRIPTION
**Pull request by furszy**
Work decoupled from #28574. Brother of #28894.

Includes two different, yet interconnected, performance and code improvements to the zap wallet transactions process.

1) As the goal of the `ZapSelectTx` function is to erase tx records that match any of the inputted hashes. There is no need to traverse the whole database record by record. We could just check if the tx exist, and remove it directly by calling `EraseTx()`.

2) Instead of performing single write operations per removed tx record, this PR batches them all within a single atomic db txn.

Moreover, these changes will enable us to consolidate all individual write operations that take place during the wallet migration process into a single db txn in the future.

### CODE REVIEW

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -2311,18 +2311,24 @@ DBErrors CWallet::LoadWallet()
     return nLoadWalletRet;
 }
 
-DBErrors CWallet::ZapSelectTx(std::vector<uint256>& vHashIn)
+void CWallet::ClearTxns(const std::vector<uint256>& tx_hashes)
```
Comment (achow101) [inline src/wallet/wallet.cpp:2314]: In 559529480fedd4181dab76f4452b5998d0cdadc3 "refactor: decouple wallet EraseTxns from ZapSelectTx"

The commit message calls this `EraseTxns`.

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -2311,18 +2311,24 @@ DBErrors CWallet::LoadWallet()
...
 {
-    AssertLockHeld(cs_wallet);
-    DBErrors nZapSelectTxRet = WalletBatch(GetDatabase()).ZapSelectTx(vHashIn);
-    for (const uint256& hash : vHashIn) {
+    for (const uint256& hash : tx_hashes) {
         const auto& it = mapWallet.find(hash);
         wtxOrdered.erase(it->second.m_it_wtxOrdered);
-        for (const auto& txin : it->second.tx->vin)
+        for (const auto& txin: it->second.tx->vin)
```
Comment (achow101) [inline src/wallet/wallet.cpp:2319]: In 559529480fedd4181dab76f4452b5998d0cdadc3 "refactor: decouple wallet EraseTxns from ZapSelectTx"

Unnecessary whitespace change.

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -3931,10 +3933,26 @@ bool CWallet::ApplyMigrationData(MigrationData& data, bilingual_str& error)
     watchonly_batch.reset(); // Flush
     // Do the removes
     if (txids_to_delete.size() > 0) {
-        if (ZapSelectTx(txids_to_delete) != DBErrors::LOAD_OK) {
```
Comment (achow101) [inline src/wallet/wallet.cpp]: In ff214259de28216a5190c358b786306985f30227 "wallet: migration, remove watch-only transactions atomically"

It seems unnecessary to be making this change here as `ZapSelectTx` is now atomic.

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (furszy) [reply src/wallet/wallet.cpp:2314]: fixed

File: src/wallet/wallet.cpp (hunk 2 above)
Comment (furszy) [reply src/wallet/wallet.cpp:2319]: fixed

File: src/wallet/wallet.cpp (hunk 3 above)
Comment (furszy) [reply src/wallet/wallet.cpp]: > In [ff21425](https://github.com/bitcoin/bitcoin/commit/ff214259de28216a5190c358b786306985f30227) "wallet: migration, remove watch-only transactions atomically"
> 
> It seems unnecessary to be making this change here as `ZapSelectTx` is now atomic.

True. This PR does not requires it. Will re-introduced it within #28574.

**Review by furszy — COMMENTED** [review]
Updated per feedback. Thanks achow. [Diff](https://github.com/bitcoin/bitcoin/compare/ff214259de28216a5190c358b786306985f30227..d26a28508d5322fe7c29b3642dcf3a17b0ea31f1).

**Review by furszy — COMMENTED** [review]
CI failure is unrelated. Relates to #29112.

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -2326,16 +2326,18 @@ void CWallet::ClearTxns(const std::vector<uint256>& tx_hashes)
...
-    // Remove transactions from the memory map
-    ClearTxns(vHashIn);
 
-    if (nZapSelectTxRet != DBErrors::LOAD_OK)
-        return nZapSelectTxRet;
+    WalletBatch batch(GetDatabase());
+    if (!batch.TxnBegin()) return DBErrors::NONCRITICAL_ERROR;
+    DBErrors db_status = batch.ZapSelectTx(vHashIn);
+    // All deletions should have been added to the db txn, otherwise we abort the process.
+    if (db_status != DBErrors::LOAD_OK) return db_status;
```
Comment (achow101) [inline src/wallet/wallet.cpp:2334]: In d26a28508d5322fe7c29b3642dcf3a17b0ea31f1 "wallet: batch ZapSelectTx db operations"

This should explicitly call `TxnAbort` before returning as otherwise we will get extraenous logs about the tx being aborted before batch is destroyed.

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -2326,16 +2326,18 @@ void CWallet::ClearTxns(const std::vector<uint256>& tx_hashes)
...
 
+    // Apply db changes and remove transactions from the memory map
+    if (!batch.TxnCommit()) return DBErrors::NONCRITICAL_ERROR;
```
Comment (achow101) [inline src/wallet/wallet.cpp:2337]: In d26a28508d5322fe7c29b3642dcf3a17b0ea31f1 "wallet: batch ZapSelectTx db operations"

I'm not sure if `NONCRITICAL_ERROR` is really the right return code. It seems like something has gone critically wrong if `TxnCommit` were to fail.

File: src/wallet/walletdb.cpp [hunk 6]
```diff
@@ -1230,88 +1230,21 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
...
+            continue;
         }
-        else if ((*it) == hash) {
-            if(!EraseTx(hash)) {
-                LogPrint(BCLog::WALLETDB, "Transaction was found for deletion but returned database error: %s\n", hash.GetHex());
-                delerror = true;
-            }
-            vTxHashOut.push_back(hash);
+        if (!EraseTx(hash)) {
+            LogPrint(BCLog::WALLETDB, "%s: Failure removing transaction: %s\n", __func__, hash.GetHex());
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/wallet/walletdb.cpp]: > Any reason to add __func__ to this log, when it previously wasn't there?

just an old bad habit. Removed. Thanks.

File: src/wallet/wallet.cpp (hunk 4 above)
Comment (furszy) [reply src/wallet/wallet.cpp]: > This should explicitly call TxnAbort before returning as otherwise we will get extraenous logs about the tx being aborted before batch is destroyed.

sure. Fixed.

File: src/wallet/wallet.cpp (hunk 5 above)
Comment (furszy) [reply src/wallet/wallet.cpp]: > I'm not sure if `NONCRITICAL_ERROR` is really the right return code. It seems like something has gone critically wrong if `TxnCommit` were to fail.

The returned error code doesn't really matter. Anything different from `LOAD_OK` is an error for all the function's callers. And they behave exactly the same for all codes.
Actually, the `DBErrors` return doesn't make sense anymore after this PR changes. The function is no longer traversing the entire db nor checking for extra, unneeded, stuff like the version inside anymore.

Based on this, I have reworked the PR cleaning up this aspect as well.

**Review by furszy — COMMENTED** [review]
Updated per feedback. Thanks Marko and achow.

**Comment by achow101** [comment]
ACK 9a3c5c8697659e34d0476103af942a4615818f4e

The new code is much simpler.

**Comment by josibake** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/28987/commits/9a3c5c8697659e34d0476103af942a4615818f4e

looks good, also `RemoveTxs` is a much better name. Might be worth updating the description to mention the rename since the description ends up in the commit log.

File: src/wallet/wallet.cpp [hunk 7]
```diff
@@ -3935,19 +3951,10 @@ bool CWallet::ApplyMigrationData(MigrationData& data, bilingual_str& error)
     watchonly_batch.reset(); // Flush
     // Do the removes
     if (txids_to_delete.size() > 0) {
-        std::vector<uint256> deleted_txids;
-        if (ZapSelectTx(txids_to_delete, deleted_txids) != DBErrors::LOAD_OK) {
-            error = _("Error: Could not delete watchonly transactions");
+        if (auto res = RemoveTxs(txids_to_delete); !res) {
+            error = _("Error: Could not delete watchonly transactions. ") + util::ErrorString(res);
```
Comment (hebasto) [inline src/wallet/wallet.cpp:3955]: nit: The trailing space could be easily overlooked and dropped during translation.

Comment (furszy) [reply src/wallet/wallet.cpp:3955]: > nit: The trailing space could be easily overlooked and dropped during translation.

@hebasto, what about creating a constant to represent the whitespace, or alternatively, implementing a function that concatenates strings with the whitespace inserted between them?
